### PR TITLE
Header flooding

### PR DESCRIPTION
**Pull request by VoidingWarranties**
This PR implements header broadcasting.

I tried to keep the changes to existing consensus code as minimal as possible. The only real changes to existing code should be:
* `validateHeader` renamed to `validateHeaderAndBlock`
* `AcceptBlock` and `threadedReceiveBlocks` broadcast blocks to some peers and headers to others, depending on their versions.

Everything else is new code and only deals with header broadcasting.

`RelayHeader` RPC mirrors the `RelayBlock` RPC used for block flooding. Every call to `RelayBlock` is also replaced with a branching broadcast that broadcasts `RelayHeader` to new peers (v0.5.2 and newer) and `RelayBlock` to old peers.

`managedAcceptHeader` mirrors `managedAcceptBlock` and does all of the header validation. However, unlike its counterpart, it cannot accept anything into the blockchain as it does not have the block and header-first blockchain downloads haven't been implemented yet. Instead, it returns `nil` if the header is valid. It is then up to the caller of `managedAcceptHeader` ( `RelayHeader`) to download the block and call either `managedAcceptBlock` or `AcceptBlock` on it.

`Send1Blk` is analogous to `SendBlocks`, but for just one block. It is used to download a block corresponding to a validated block header. Peers can request blocks from other peers by the block id. The requesting peer then calls `AcceptBlock` on the received block.

I tried to keep the commits clean and separated, so going through the diffs commit by commit should be easy.

### CODE REVIEW

File: modules/consensus/synchronize.go [hunk 1]
```diff
@@ -245,16 +257,106 @@ func (cs *ConsensusSet) relayBlock(conn modules.PeerConn) error {
...
+// the function it returns calls the exported method AcceptBlock.
+func (cs *ConsensusSet) threadedReceiveBlock(id types.BlockID) modules.RPCFunc {
+	managedFN := func(conn modules.PeerConn) error {
+		if err := encoding.WriteObject(conn, id); err != nil {
+			return err
+		}
+		var block types.Block
+		if err := encoding.ReadObject(conn, &block, types.BlockSizeLimit); err != nil {
+			return err
+		}
```
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: If `Send1Blk` errors out (if, for example the peer doesn't have the requested block), this call will block indefinitely. This would be fixed by #988

File: modules/consensus/accept_test.go [hunk 2]
```diff
@@ -417,6 +612,16 @@ func TestOrphanHandling(t *testing.T) {
 	if err != errOrphan {
 		t.Fatalf("expected %v, got %v", errOrphan, err)
 	}
+	// Try submitting an orphan block's header.
+	// TODO: why is the test above duplicated?
```
Comment (VoidingWarranties) [inline modules/consensus/accept_test.go]: No harm in duplicating the test, but was it purposeful?

File: modules/consensus/synchronize.go [hunk 3]
```diff
@@ -245,16 +257,106 @@ func (cs *ConsensusSet) relayBlock(conn modules.PeerConn) error {
...
+		// TODO: are these sanity checks necessary?
+		pathID, err := getPath(tx, pb.Height)
+		if err != nil {
+			return err
+		}
+		if pathID != pb.Block.ID() {
+			return errors.New("pathID and processed block's ID do not match")
+		}
```
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: What is the reason for these sanity checks? (they also exist in `sendBlocks`)

File: modules/consensus/synchronize_test.go [hunk 4]
```diff
@@ -311,3 +316,488 @@ func TestSendBlocksBroadcastsOnce(t *testing.T) {
...
+				if err != nil {
+					t.Error(err)
+				}
+				fnDone <- struct{}{}
+			},
+			errWant: errNilItem,
+			msg:     "expected send1Blk to error with a nonexistent block id",
+		},
+		// TODO: Other sanity checks? e.g. test if getPath returns an error or pathID != pb.Block.ID().
+		// TODO: Test with an otherwise failing db?
```
Comment (VoidingWarranties) [inline modules/consensus/synchronize_test.go]: Wasn't sure how to easily test these.

File: modules/consensus/synchronize_test.go [hunk 5]
```diff
@@ -311,3 +316,488 @@ func TestSendBlocksBroadcastsOnce(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cst2.Close()
+
+	err = cst1.cs.gateway.Connect(cst2.cs.gateway.Address())
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = cst2.cs.gateway.Connect(cst1.cs.gateway.Address()) // TODO: why is this Connect call necessary? Shouldn't they connect to eachother with one Connect?
```
Comment (VoidingWarranties) [inline modules/consensus/synchronize_test.go]: A gateway receiving a connection request should connect to the requesting gateway, but this doesn't happen here. Why?

File: modules/consensus/accept.go [hunk 6]
```diff
@@ -55,6 +56,73 @@ func (cs *ConsensusSet) validateHeader(tx dbTx, b types.Block) error {
...
+	if err != nil {
+		return err
+	}
+
+	// Check that the target of the new block is sufficient.
+	if !checkHeaderTarget(h, parent.ChildTarget) {
+		return modules.ErrBlockUnsolved
+	}
+
+	// TODO: check if the block is a non extending block.
```
Comment (VoidingWarranties) [inline modules/consensus/accept.go]: I _think_ this can be determined with just a header, but the current implementation requires a `processedBlock` and we don't want processedBlocks if we don't have the whole block.

File: modules/consensus/accept.go [hunk 7]
```diff
@@ -154,6 +222,7 @@ func (cs *ConsensusSet) managedAcceptBlock(b types.Block) error {
 			if err == errFutureTimestamp {
 				go func() {
 					time.Sleep(time.Duration(b.Timestamp-(types.CurrentTimestamp()+types.FutureThreshold)) * time.Second)
+					// TODO: log error returned by AcceptBlock if non-nill?
```
Comment (VoidingWarranties) [inline modules/consensus/accept.go]: Would be easy if the consensus had logging, but since it doesn't yet, adding it should be another PR.

**Comment by DavidVorick** [comment]
should change rpc names to 8 chars, enforced by panics at runtime

File: modules/consensus/accept.go (hunk 6 above)
Comment (DavidVorick) [inline modules/consensus/accept.go]: you can determine that with just a header, though you don't want to trust it because you don't actually know if the data for the block exists. And you don't want to skip downloading the block if it's non-extending because you need to be able to get reorgs.

A check here doesn't make sense until we're doing headers-first downloads.

File: modules/consensus/accept_test.go (hunk 2 above)
Comment (DavidVorick) [inline modules/consensus/accept_test.go]: I remember that there was a good reason, but I forget why.

File: types/block.go [hunk 8]
```diff
@@ -99,6 +103,13 @@ func (b Block) ID() BlockID {
 	return BlockID(crypto.HashObject(b.Header()))
 }
 
+// ID returns the ID of a Block, which is calculated by hashing the
+// concatenation of the block's parent's ID, nonce, and the result of the
+// b.MerkleRoot().
+func (h BlockHeader) ID() BlockID {
```
Comment (DavidVorick) [inline types/block.go]: The comment here is confusing. While correct, the logic just hashes the header, which doesn't seem to match the comment.

File: modules/consensus/synchronize.go (hunk 1 above)
Comment (DavidVorick) [inline modules/consensus/synchronize.go]: that's bad. This should be tolerant and should not rely on the gateway forcing timeouts.

File: modules/consensus/synchronize.go (hunk 3 above)
Comment (DavidVorick) [inline modules/consensus/synchronize.go]: This was to help guarantee that the path was being constructed correctly, and that the processed blocks were being saved correctly (instead of overwriting eachother). That's pretty well tested at this point, the sanity check can be safely removed I think.

File: modules/consensus/synchronize_test.go (hunk 5 above)
Comment (DavidVorick) [inline modules/consensus/synchronize_test.go]: does it have something to do with inbound/outbound peers, or some level of mocking that's happening somewhere? Perhaps it's a bug.

File: types/block_test.go [hunk 9]
```diff
@@ -113,6 +113,44 @@ func TestBlockID(t *testing.T) {
 	}
 }
 
+// TestHeaderID probes the ID function of the BlockHeader type.
+func TestHeaderID(t *testing.T) {
```
Comment (DavidVorick) [inline types/block_test.go]: might make sense to throw a sanity check somewhere in the consensus code to make sure that the block id and the header id always yield the same result. Then we'd get to verify it when running a debugging instance of 'siad', because it would check across a wide range of block types.

File: types/block.go [hunk 10]
```diff
@@ -13,6 +13,10 @@ import (
 	"github.com/NebulousLabs/Sia/encoding"
 )
 
+const (
+	BlockHeaderSize = 80 // 32 (ParentID) + 8 (Nonce) + 8 (Timestamp) + 32 (MerkleRoot)
```
Comment (DavidVorick) [inline types/block.go]: needs a docstring

File: modules/gateway/rpc.go [hunk 11]
```diff
@@ -114,6 +114,7 @@ func (g *Gateway) threadedHandleConn(conn modules.PeerConn) {
 		g.log.Printf("WARN: incoming conn %v requested unknown RPC \"%v\"", conn.RemoteAddr(), id)
 		return
 	}
+	g.log.Printf("INFO: incoming conn %v requested RPC \"%v\"", conn.RemoteAddr(), id)
```
Comment (DavidVorick) [inline modules/gateway/rpc.go]: if you put this here the gateway.log is going to be too big. It's already gigabytes large for some longer-running nodes. this log statement can't go in a normal log (Luke likes the idea of a rolling log, though I'm not sure that's the best approach - it's probably good enough in this case to just not log at all)

Comment (DavidVorick) [inline modules/gateway/rpc.go]: *if you really want the log, have it write the log in debug mode. bitcoind will make stupid-large logs (sometimes running into the hundreds of GBs) if you run it in debug/verbose mode. It's probably okay for us to take a similar approach. Non-debug logs only log when something unanticipated goes wrong (so, failed RPCs and random network drops would not merit logs), and debug logs just dump data.

File: modules/consensus/accept.go [hunk 12]
```diff
@@ -55,6 +56,73 @@ func (cs *ConsensusSet) validateHeader(tx dbTx, b types.Block) error {
...
+	// the extreme future arrives, this block will no longer be a part of the
+	// longest fork because it will have been ignored by all of the miners.
+	if h.Timestamp > types.CurrentTimestamp()+types.ExtremeFutureThreshold {
+		return errExtremeFutureTimestamp
+	}
+
+	// Check if the block is in the near future, but too far to be acceptable.
+	if h.Timestamp > types.CurrentTimestamp()+types.FutureThreshold {
+		return errFutureTimestamp
+	}
```
Comment (DavidVorick) [inline modules/consensus/accept.go]: no need to have a check for errFutureTimestamp, that's logic that should be handled in acceptBlock only. You may want to add a comment explaining why. The reason is that you want to have the future blocks in memory, so that you can accept them the millisecond that they are no longer too far in the future to be acceptable. Maybe they will have reorged, maybe not. But you want the host to go chase down the block if it finds a header, even if it's a 'future' block.

File: modules/consensus/accept.go (hunk 7 above)
Comment (DavidVorick) [inline modules/consensus/accept.go]: in that case leaving a TODO is fine.

File: modules/consensus/accept.go [hunk 13]
```diff
@@ -201,8 +270,38 @@ func (cs *ConsensusSet) AcceptBlock(b types.Block) error {
...
+		// Do not accept a header if the database is inconsistent.
+		if inconsistencyDetected(tx) {
+			return errInconsistentSet
+		}
+
+		// Do some relatively inexpensive checks to validate the header
+		return cs.validateHeader(boltTxWrapper{tx}, h)
+	})
+	return err
+}
```
Comment (DavidVorick) [inline modules/consensus/accept.go]: why do you need a managed version of AcceptHeader? The reason you need a managed version of accept block is because the lock has to be demoted partway through the function, which means you can't wrap it in a lock. AcceptHeader on the other hand doesn't push any updates to the modules (nor does it even modify the consensus set), so you don't need to worry about demotion. I don't think this function needs to be here, nor does it need to be managed.

File: modules/consensus/synchronize.go [hunk 14]
```diff
@@ -96,8 +98,17 @@ func (cs *ConsensusSet) threadedReceiveBlocks(conn modules.PeerConn) error {
...
+			var relayBlockPeers, relayHeaderPeers []modules.Peer
+			for _, p := range cs.gateway.Peers() {
+				if build.VersionCmp(p.Version, "0.5.1") <= 0 {
+					relayBlockPeers = append(relayBlockPeers, p)
+				} else {
+					relayHeaderPeers = append(relayHeaderPeers, p)
+				}
+			}
+			go cs.gateway.Broadcast("RelayBlock", currentBlock, relayBlockPeers)
+			go cs.gateway.Broadcast("RelayHeader", currentBlock.Header(), relayHeaderPeers)
```
Comment (DavidVorick) [inline modules/consensus/synchronize.go]: might be worth abstracting this code into a helper function, since it's written in a few places.

File: modules/consensus/synchronize.go [hunk 15]
```diff
@@ -246,16 +257,106 @@ func (cs *ConsensusSet) relayBlock(conn modules.PeerConn) error {
...
+	switch err {
+	case errOrphan:
+		// If the header is an orphan, try to find the parents.
+		// TODO: log error returned if non-nill?
 		go cs.gateway.RPC(modules.NetAddress(conn.RemoteAddr().String()), "SendBlocks", cs.threadedReceiveBlocks)
+	case errFutureTimestamp:
+		// Ignore errFutureTimestamp so that the block is downloaded and passed to
+		// managedAcceptHeader, which will spawn a go thread that waits until the
+		// timestamp would be valid before trying to accept it again.
+		fallthrough
```
Comment (DavidVorick) [inline modules/consensus/synchronize.go]: since we're removing errFutureTimestamp from the set of errors that acceptHeader can return, there's no need for this case or fallthrough.

File: modules/consensus/synchronize.go [hunk 16]
```diff
@@ -246,16 +257,106 @@ func (cs *ConsensusSet) relayBlock(conn modules.PeerConn) error {
...
 	return nil
 }
+
+// threadedReceiveBlock takes a block id and returns an RPCFunc that requests
+// that block and then calls AcceptBlock on it. The returned function should be
+// used as the calling end of the Send1Blk RPC. Note that although the function
+// itself does not do any locking, it is still prefixed with "threaded" because
+// the function it returns calls the exported method AcceptBlock.
+func (cs *ConsensusSet) threadedReceiveBlock(id types.BlockID) modules.RPCFunc {
+	managedFN := func(conn modules.PeerConn) error {
```
Comment (DavidVorick) [inline modules/consensus/synchronize.go]: `return func() error { ... }`

**Comment by DavidVorick** [comment]
I haven't looked at the testing yet, otherwise I've finished making this round of comments.

File: modules/consensus/synchronize.go (hunk 1 above)
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: `SendBlocks` and `threadedReceiveBlocks` is also susceptible to this. The solution is simple enough I think. Something like:
```go
// send1Blk
err = cs.db.View(func(tx *bolt.Tx) error {
// ...
})
if err != nil {
	encoding.WriteObject(conn, types.Block{})
	return err
}
// ...

// sendBlocks
err = cs.db.View(func(tx *bolt.Tx) error {
// ...
})
if err != nil {
	encoding.WriteObject(conn, []types.Block)
	encoding.WriteObject(conn, false)
	return err
}
// ...
```
But then the errors from `encoding.WriteObject` are lost.

File: modules/consensus/synchronize.go (hunk 3 above)
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: I suppose you could request a block using `Send1Blk` that isn't in the longest chain. In that case the `pathID != pb.Block.ID()` would be true.

File: modules/consensus/synchronize.go (hunk 1 above)
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: Alternatively we could concatenate the errors like you did in #989 with `composeErrors`.

File: types/block_test.go (hunk 9 above)
Comment (VoidingWarranties) [inline types/block_test.go]: Where would be a good place for the sanity check? This test does verify that `BlockHeader.ID()` and `Block.ID()` match for the test blocks FWIW.

Comment (DavidVorick) [inline types/block_test.go]: I'd probably throw the sanity check in validateHeaderAndBlock

all the test blocks is one thing, but all of the blocks in the whole consensus set is another level of assurance, especially because it's a sanity check that'll be run for every block that gets mined in all the integration tests as well

File: modules/consensus/synchronize.go (hunk 14 above)
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: Since RelayBlock will eventually be removed once enough peers upgrade I think it's okay to not use a helper function. This code is only duplicated in two places.

File: modules/consensus/synchronize.go [hunk 17]
```diff
@@ -247,12 +256,16 @@ func (cs *ConsensusSet) relayBlock(conn modules.PeerConn) error {
...
 	err = cs.AcceptBlock(b)
 	if err == errOrphan {
 		// If the block is an orphan, try to find the parents. The block
 		// received from the peer is discarded and will be downloaded again if
 		// the parent is found.
+		//
+		// TODO: log error returned if non-nill? OR we could not use a goroutine and
+		// just return the result of the RPC call. Is it back practice to call an RPC
+		// from inside another RPC?
 		go cs.gateway.RPC(modules.NetAddress(conn.RemoteAddr().String()), "SendBlocks", cs.threadedReceiveBlocks)
```
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: Could this line be
```go
return cs.gateway.RPC(modules.NetAddress(conn.RemoteAddr().String()), "SendBlocks", cs.threadedReceiveBlocks)
```
?

File: modules/consensus/synchronize.go [hunk 18]
```diff
@@ -260,3 +273,96 @@ func (cs *ConsensusSet) relayBlock(conn modules.PeerConn) error {
...
+	})
+	cs.mu.RUnlock()
+	if err == errOrphan {
+		// If the header is an orphan, try to find the parents.
+		//
+		// TODO: log error returned if non-nill? OR we could not use a goroutine and
+		// just return the result of the RPC call. Is it back practice to call an RPC
+		// from inside another RPC?
+		go cs.gateway.RPC(modules.NetAddress(conn.RemoteAddr().String()), "SendBlocks", cs.threadedReceiveBlocks)
+		return nil
```
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: Same for this. Could we return the result of the RPC call instead of spawning a goroutine?

**Comment by VoidingWarranties** [comment]
Mostly done except for the few comments I just made.

Also it look like Github threw out all the comments @DavidVorick made on the individual commits because I rebased. :( I guess we should only make comments on the master diff.

File: modules/consensus/accept.go [hunk 19]
```diff
@@ -140,20 +210,31 @@ func (cs *ConsensusSet) managedAcceptBlock(b types.Block) error {
...
+		err := cs.validateHeaderAndBlock(boltTxWrapper{tx}, b)
 		if err != nil {
 			// If the block is in the near future, but too far to be acceptable, then
 			// save the block and add it to the consensus set after it is no longer
 			// too far in the future.
+			//
+			// TODO: an attacker could mine many blocks off the genesis block all in the
+			// future and we would spawn a goroutine per each block. To fix this, either
+			// ban peers that send lots of future blocks OR stop spawning goroutines
+			// after we are already waiting on a large number of future blocks.
```
Comment (DavidVorick) [inline modules/consensus/accept.go]: probably both. You want a hard-limit on all resources that you use, I'd never advocate having a system that can spin up unlimited goroutines.

File: modules/consensus/accept.go [hunk 20]
```diff
@@ -140,20 +210,31 @@ func (cs *ConsensusSet) managedAcceptBlock(b types.Block) error {
...
+			//
+			// TODO: an attacker could broadcast a future block many times and we would
+			// spawn a goroutine for each broadcast. To fix this we should create a
+			// cache of future blocks, like we already do for DoS blocks, and only spawn
+			// a goroutine if we haven't already spawned one for that block.
```
Comment (DavidVorick) [inline modules/consensus/accept.go]: and then start evicting blocks once there are 50. We'd evict the ones that are furthest in the future.

File: modules/consensus/accept_test.go [hunk 21]
```diff
@@ -286,6 +311,181 @@ func TestUnitValidateHeader(t *testing.T) {
...
+	var h types.BlockHeader
+
+	tests := []struct {
+		target   types.Target
+		expected bool
+		msg      string
+	}{
+		{types.RootDepth, true, "checkHeaderTarget failed for a low target"},
+		{types.Target{}, false, "checkHeaderTarget passed for a high target"},
+		{types.Target(h.ID()), true, "checkHeaderTarget failed for a same target"},
```
Comment (DavidVorick) [inline modules/consensus/accept_test.go]: huh. I've always done this by manually mining a target that either did or didn't fit the criteria. The testing target is {128}, which means you have an exactly 50% chance of finding a valid header each attempt. Not too painful to write a loop that guarantees the correct outcome, it's cryptographically guaranteed to complete within 128 iterations.

The way you've done it is fine as well

File: modules/consensus/synchronize.go [hunk 22]
```diff
@@ -96,8 +96,17 @@ func (cs *ConsensusSet) threadedReceiveBlocks(conn modules.PeerConn) error {
...
 			currentBlock := cs.CurrentBlock()
-			peers := cs.gateway.Peers()
-			go cs.gateway.Broadcast("RelayBlock", currentBlock, peers)
+			// Broadcast the block to all peers <= v0.5.1 and block header to all peers > v0.5.1
+			var relayBlockPeers, relayHeaderPeers []modules.Peer
+			for _, p := range cs.gateway.Peers() {
+				if build.VersionCmp(p.Version, "0.5.1") <= 0 {
+					relayBlockPeers = append(relayBlockPeers, p)
+				} else {
+					relayHeaderPeers = append(relayHeaderPeers, p)
```
Comment (DavidVorick) [inline modules/consensus/synchronize.go]: no coverage on this line of code

File: modules/consensus/synchronize.go (hunk 17 above)
Comment (DavidVorick) [inline modules/consensus/synchronize.go]: I don't object to that.

Comment (DavidVorick) [inline modules/consensus/synchronize.go]: @lukechampine

File: modules/consensus/synchronize.go (hunk 18 above)
Comment (DavidVorick) [inline modules/consensus/synchronize.go]: same for this

File: modules/consensus/synchronize.go [hunk 23]
```diff
@@ -260,3 +273,96 @@ func (cs *ConsensusSet) relayBlock(conn modules.PeerConn) error {
...
+	} else if err != nil {
+		return err
+	}
+	// If the header is valid and extends the heaviest chain, fetch, accept it,
+	// and broadcast it.
+	//
+	// TODO: log error returned if non-nill? OR we could not use a goroutine and
+	// just return the result of the RPC call. Is it back practice to call an RPC
+	// from inside another RPC?
+	go cs.gateway.RPC(modules.NetAddress(conn.RemoteAddr().String()), "BlockID", cs.threadedReceiveBlock(h.ID()))
```
Comment (DavidVorick) [inline modules/consensus/synchronize.go]: and this

File: types/block_test.go (hunk 9 above)
Comment (DavidVorick) [inline types/block_test.go]: actually, I see now that the definition of Block.ID is `crypto.HashObject(b.Header())`

sanity check is probably not needed

File: types/block.go [hunk 24]
```diff
@@ -99,6 +105,11 @@ func (b Block) ID() BlockID {
 	return BlockID(crypto.HashObject(b.Header()))
 }
 
+// ID returns the ID of a Block, which is calculated by hashing the header.
+func (h BlockHeader) ID() BlockID {
```
Comment (DavidVorick) [inline types/block.go]: needs to be sorted by object. You've got a blockHeader function between a bunch of functions that operate on a block. I'd probably move it to above the block, though if you want to stick with the alphabetic convention you should move it below all the block functions.

Either one is fine with me, but I'd rather it not be in the middle of a bunch of block methods

File: modules/consensus/synchronize.go [hunk 25]
```diff
@@ -260,3 +273,96 @@ func (cs *ConsensusSet) relayBlock(conn modules.PeerConn) error {
...
+		return err
+	}
+
+	// Start verification inside of a bolt View tx.
+	cs.mu.RLock()
+	err = cs.db.View(func(tx *bolt.Tx) error {
+		// Do not accept a header if the database is inconsistent.
+		if inconsistencyDetected(tx) {
+			return errInconsistentSet
+		}
```
Comment (DavidVorick) [inline modules/consensus/synchronize.go]: I don't think you need this. It's more of a startup sanity check, and was a bigger issue before we were atomically modifying the database.

File: modules/consensus/synchronize.go [hunk 26]
```diff
@@ -260,3 +273,96 @@ func (cs *ConsensusSet) relayBlock(conn modules.PeerConn) error {
...
+		// TODO: log error returned if non-nill? OR we could not use a goroutine and
+		// just return the result of the RPC call. Is it back practice to call an RPC
+		// from inside another RPC?
+		go cs.gateway.RPC(modules.NetAddress(conn.RemoteAddr().String()), "SendBlocks", cs.threadedReceiveBlocks)
+		return nil
+	} else if err != nil {
+		return err
+	}
+	// If the header is valid and extends the heaviest chain, fetch, accept it,
+	// and broadcast it.
```
Comment (DavidVorick) [inline modules/consensus/synchronize.go]: too many  steps in the comment.

`If the header is valid, go get the corresponding block for full validation`.

File: modules/consensus/synchronize.go [hunk 27]
```diff
@@ -260,3 +273,96 @@ func (cs *ConsensusSet) relayBlock(conn modules.PeerConn) error {
...
+	//
+	// TODO: log error returned if non-nill? OR we could not use a goroutine and
+	// just return the result of the RPC call. Is it back practice to call an RPC
+	// from inside another RPC?
+	go cs.gateway.RPC(modules.NetAddress(conn.RemoteAddr().String()), "BlockID", cs.threadedReceiveBlock(h.ID()))
+	return nil
+}
+
+// rpcBlockID is an RPC that sends the requested block to the requesting peer.
+func (cs *ConsensusSet) rpcBlockID(conn modules.PeerConn) error {
```
Comment (DavidVorick) [inline modules/consensus/synchronize.go]: I don't like this name, makes it seem like you are fetching a blockID. Maybe just `rpcBlock` is good enough.

File: modules/consensus/synchronize.go (hunk 1 above)
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: Actually this isn't a problem. When `Send1Blk` errors out the gateway closes the connection, which should cause any reads or writes on the other end to error out as well.

File: modules/consensus/synchronize.go (hunk 27 above)
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: What about `SendBlk` instead of `BlockID` or `Send1Blk`? I don't think `Block` makes it clear that not only are we fetching the block, since we are also accepting it into the consensus set?

File: modules/consensus/synchronize.go (hunk 26 above)
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: But all of those things are done. I want to make it very clear that the RPC to fetch the corresponding block to a block header will do all of those things.

File: modules/consensus/synchronize.go (hunk 22 above)
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: This code is just copy-pasted logic from the similar change we made to broadcasting `RelayTransactionSet` vs `RelayTransaction` which is already tested. If we're going to be removing this line in the future (once enough peers have upgraded) I'd rather not spend more time testing it.

File: modules/consensus/synchronize.go (hunk 17 above)
Comment (VoidingWarranties) [inline modules/consensus/synchronize.go]: It becomes problematic if cyclical RPC calls are possible. In fact this chain of RPC calls is cyclical.

Also it prevents us from setting reasonable timeouts. If RPC "a" sets a timeout of 5 minutes, and RPC "b" calls RPC "a", there is no way for RPC "b" to know that it should extend its timeout by 5 minutes.

So I'll keep them called in goroutines for now and add logging later.

**Comment by VoidingWarranties** [comment]
Not ready to merge yet, adding more integration tests.

**Comment by VoidingWarranties** [comment]
Spun up a test node and it's working well. Should be good to merge.

**Comment by DavidVorick** [comment]
I remember estimating that this would be a simple PR. and initially, it was. I'm glad we took the time to make it quality though.

**Comment by DavidVorick** [comment]
As of this PR, you are going to be primarily in charge of the peer network, the gateway code, and the synchronization process. You've had to do a deep enough dive to figure all this out that you probably are more familiar with the code than either Luke or myself.

Congrats :)
